core/src/main/java/tc/oc/pgm/util/xml/parsers/Builder.java

```diff
@@ -0,0 +1,114 @@
+package tc.oc.pgm.util.xml.parsers;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import org.jdom2.Element;
+import org.jetbrains.annotations.Nullable;
+import tc.oc.pgm.util.function.ThrowingSupplier;
+import tc.oc.pgm.util.xml.InvalidXMLException;
+import tc.oc.pgm.util.xml.Node;
+
+public abstract class Builder<T, B extends Builder<T, B>> {
+  protected final @Nullable Element el;
+  protected final String[] prop;
+
+  protected List<Validator<T>> validators;
+  protected boolean attr;
+  protected boolean child;
+  protected boolean self;
+
+  public Builder(@Nullable Element el, String... prop) {
+    this.el = el;
+    this.prop = prop;
+    this.attr = prop.length > 0;
+    this.child = prop.length > 0;
+    this.self = prop.length == 0;
+  }
+
+  /** Make it so this parser will only run on attributes, ignoring child elements */
+  public B attr() {
+    assert prop.length != 0;
+    attr = true;
+    child = false;
+    return getThis();
```

Comment:
I don't understand the purpose of having `getThis()` instead of returning `this`.

Anyways this is not a blocking conern

Reply:
Extended builders

If you return `this` in the parent `Builder`, you can't know what you're returning, because a parent can't know what it's subclass is

Reply:
Ah, that makes sense